Add module-level version of constant-propagation pass.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24285 Add module-level version of constant-propagation pass.**

Summary:
Generally most of our passes should work on module level rather than on a graph
level, even though it might make sense for them to have graph version as well.
Constant propagation is a pretty popular pass that can be useful in many
scenarious as just a cleanup, and for it it definitely makes sense to have a
module-level version.

Differential Revision: [D16824525](https://our.internmc.facebook.com/intern/diff/D16824525)

Commit:
Add module-level version of constant-propagation pass.

Summary:
Generally most of our passes should work on module level rather than on a graph
level, even though it might make sense for them to have graph version as well.
Constant propagation is a pretty popular pass that can be useful in many
scenarious as just a cleanup, and for it it definitely makes sense to have a
module-level version.

Commit:
Update on "Add module-level version of constant-propagation pass."


Summary:
Generally most of our passes should work on module level rather than on a graph
level, even though it might make sense for them to have graph version as well.
Constant propagation is a pretty popular pass that can be useful in many
scenarious as just a cleanup, and for it it definitely makes sense to have a
module-level version.

suo commented:
Who is the envisioned user of this? I'm a little skeptical, as it makes constant propagation somehow different from every other pass. Keeping passes local to graphs is good, since that's the "real" unit of optimization and execution. I don't want people to start shoving in state to coordinate pass behavior just because the graphs in question belong to the same module.

Maybe it would be better to provide a generic function that says: "clone this module, then run this lambda over all the methods". This keeps const prop as a pure graph transformation and is more extensible anyway.

That said, don't feel too strongly about it.

ZolotukhinM commented:
> Who is the envisioned user of this? I'm a little skeptical, as it makes constant propagation somehow different from every other pass. Keeping passes local to graphs is good, since that's the "real" unit of optimization and execution. I don't want people to start shoving in state to coordinate pass behavior just because the graphs in question belong to the same module.
> 
> Maybe it would be better to provide a generic function that says: "clone this module, then run this lambda over all the methods". This keeps const prop as a pure graph transformation and is more extensible anyway.
> 
> That said, I don't feel too strongly about it.

The soonest user of this will be quantization: once we script a module before invoking quantization passes (insert observers et al.) we would like to cleanup the IR. All the quantization passes would operate on module scope (they need to access module attributes/parameters and be able to walk through calls to other submodules), so while it is totally doable to run the current const-prop even if it operates on graphs, it seemed reasonable to have a module version too.

Const-prop is the only one for now, but my understanding was that we were actually going to make more of them to operate on module level at some point. I might be mistaken though :) The lambda-over-all-methods approach seems good, but I'd expect a standard cleanup to be more easily accessible.

I wonder what @zdevito thinks about this?

zdevito commented:
My first impression is a similar skepticism to @suo. This breaks a lot of assumptions, primarily that Modules are "just classes". It also would assume that a module is a singleton class. That won't be true in a week or two.

Commit:
Change to in-place version. on "Add module-level version of constant-propagation pass."


Summary:
Generally most of our passes should work on module level rather than on a graph
level, even though it might make sense for them to have graph version as well.
Constant propagation is a pretty popular pass that can be useful in many
scenarious as just a cleanup, and for it it definitely makes sense to have a
module-level version.

Differential Revision: [D16824525](https://our.internmc.facebook.com/intern/diff/D16824525)
